raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientReconfigTest.java

```diff
@@ -0,0 +1,271 @@
+    public void testLeaderWritesBootstrapRecords(boolean withKip853Rpc) throws Exception {
+        int localId = 0;
+        int follower = 1;
+        int observer = 2;
+        Uuid localDirectoryId = Uuid.randomUuid();
+        Uuid followerDirectoryId = Uuid.randomUuid();
+        Uuid observerDirectoryId = Uuid.randomUuid();
+        Set<Integer> voterIds = new HashSet<>(Arrays.asList(localId, follower));
+        Set<ReplicaKey> voters = new HashSet<>(Arrays.asList(
+            ReplicaKey.of(localId, localDirectoryId),
+            ReplicaKey.of(follower, followerDirectoryId)));
+        VoterSet voterSet = VoterSetTest.voterSet(voters.stream());
+        short kRaftVersion = 1;
+        int epoch = 0;
+
+        RaftClientTestContext context = new RaftClientTestContext.Builder(OptionalInt.of(localId), voterIds, localDirectoryId, kRaftVersion)
+            .withKip853Rpc(withKip853Rpc)
+            .withBootstrapSnapshot(Optional.of(voterSet), kRaftVersion)
+            .appendToLog(epoch, Collections.singletonList("first_record"))
+            .withUnknownLeader(epoch)
+            .build();
+
+        // check the bootstrap snapshot exists and contains the expected records
+        assertEquals(BOOTSTRAP_SNAPSHOT_ID, context.log.latestSnapshotId().get());
+        RawSnapshotReader snapshot = context.log.latestSnapshot().get();
+        List<String> expectedBootstrapRecords = new ArrayList<>();
+        expectedBootstrapRecords.add(new SnapshotHeaderRecord()
+            .setVersion((short) 0)
+            .setLastContainedLogTimestamp(0).toString());
+        expectedBootstrapRecords.add(new KRaftVersionRecord()
+            .setVersion(ControlRecordUtils.KRAFT_VERSION_CURRENT_VERSION)
+            .setKRaftVersion((short) 1).toString());
+        expectedBootstrapRecords.add(voterSet.toVotersRecord(ControlRecordUtils.KRAFT_VOTERS_CURRENT_VERSION).toString());
+        expectedBootstrapRecords.add(new SnapshotFooterRecord().setVersion((short) 0).toString());
+        SnapshotWriterReaderTest.assertControlSnapshot(Collections.singletonList(expectedBootstrapRecords), snapshot);
+
+        // check if leader writes 3 bootstrap records to log after first record
+        context.becomeLeader();
+        Records records = context.log.read(1, Isolation.UNCOMMITTED).records;
+        RecordBatch batch = records.batches().iterator().next();
+        assertTrue(batch.isControlBatch());
+        Iterator<Record> recordIterator = batch.iterator();
+        Record record = recordIterator.next();
+        RaftClientTestContext.verifyLeaderChangeMessage(localId, Arrays.asList(localId, follower),
+            Arrays.asList(localId, follower), record.key(), record.value());
+        record = recordIterator.next();
+        RaftClientTestContext.verifyKRaftVersionRecord(kRaftVersion, record.key(), record.value());
+        record = recordIterator.next();
+        RaftClientTestContext.verifyVotersRecord(voterIds, record.key(), record.value());
```

Comment:
Let's not increase the scope of `RaftClientTestContext` unless it is strictly necessary. Since these are `public static` method I doubt you need them in `RaftClientTestContext`. You can declare those methods as static in this type instead.